File: forms/outpatient-outcome/back-end-with-loco/src/engine/prem_domain.rs
use super::types::{AssessmentData, DomainGrade, FiredRule};

pub struct PremResult {
    pub grade: DomainGrade,
    pub rules: Vec<FiredRule>,
}

/// Grade the PREM domain from the NHS Friends and Family Test (FFT) response.
///
/// extremely_likely    → A (Very Good)
/// likely              → B (Good)
/// neither             → C (Neither good nor poor)
/// unlikely            → D (Poor)
/// extremely_unlikely  → E (Very Poor)
/// dont_know / ''      → '' (data quality flag — no grade assigned)
pub fn grade_prem(data: &AssessmentData) -> PremResult {
    let fft = data.prem_fft.fft_response.as_str();
    let mut rules: Vec<FiredRule> = Vec::new();
    let grade: DomainGrade = match fft {
        "extremely_likely" => {
            rules.push(FiredRule {
                id: "PREM-001".to_string(),
                domain: "PREM".to_string(),
                description: "FFT response: Extremely Likely (Very Good)".to_string(),
                grade: "A".to_string(),
            });
            "A".to_string()
        }
        "likely" => {
            rules.push(FiredRule {
                id: "PREM-002".to_string(),
                domain: "PREM".to_string(),
                description: "FFT response: Likely (Good)".to_string(),
                grade: "B".to_string(),
            });
            "B".to_string()
        }
        "neither" => {
            rules.push(FiredRule {
                id: "PREM-003".to_string(),
                domain: "PREM".to_string(),
                description: "FFT response: Neither good nor poor".to_string(),
                grade: "C".to_string(),
            });
            "C".to_string()
        }
        "unlikely" => {
            rules.push(FiredRule {
                id: "PREM-004".to_string(),
                domain: "PREM".to_string(),
                description: "FFT response: Unlikely (Poor)".to_string(),
                grade: "D".to_string(),
            });
            "D".to_string()
        }
        "extremely_unlikely" => {
            rules.push(FiredRule {
                id: "PREM-005".to_string(),
                domain: "PREM".to_string(),
                description: "FFT response: Extremely Unlikely (Very Poor)".to_string(),
                grade: "E".to_string(),
            });
            "E".to_string()
        }
        _ => String::new(),
    };
    PremResult { grade, rules }
}

File: forms/outpatient-outcome/front-end-with-svelte/src/lib/engine/prem-domain.ts
import type { AssessmentData, DomainGrade, FiredRule } from './types';

/**
 * Grade the PREM domain from the NHS Friends and Family Test (FFT) response.
 *
 * FFT maps to the OOCG grade as follows (NHS FFT scale):
 * - extremely_likely  → A (Very Good)
 * - likely            → B (Good)
 * - neither           → C (Neither good nor poor)
 * - unlikely          → D (Poor)
 * - extremely_unlikely → E (Very Poor)
 * - dont_know / ''    → '' (data quality flag — no grade assigned)
 */
export function gradePREM(data: AssessmentData): { grade: DomainGrade; rules: FiredRule[] } {
	const fft = data.premFft.fftResponse;
	const rules: FiredRule[] = [];

	let grade: DomainGrade = '';

	switch (fft) {
		case 'extremely_likely':
			grade = 'A';
			rules.push({
				id: 'PREM-001',
				domain: 'PREM',
				description: 'FFT response: Extremely Likely (Very Good)',
				grade: 'A'
			});
			break;
		case 'likely':
			grade = 'B';
			rules.push({
				id: 'PREM-002',
				domain: 'PREM',
				description: 'FFT response: Likely (Good)',
				grade: 'B'
			});
			break;
		case 'neither':
			grade = 'C';
			rules.push({
				id: 'PREM-003',
				domain: 'PREM',
				description: 'FFT response: Neither good nor poor',
				grade: 'C'
			});
			break;
		case 'unlikely':
			grade = 'D';
			rules.push({
				id: 'PREM-004',
				domain: 'PREM',
				description: 'FFT response: Unlikely (Poor)',
				grade: 'D'
			});
			break;
		case 'extremely_unlikely':
			grade = 'E';
			rules.push({
				id: 'PREM-005',
				domain: 'PREM',
				description: 'FFT response: Extremely Unlikely (Very Poor)',
				grade: 'E'
			});
			break;
		default:
			// dont_know or '' → data-quality issue; no grade
			grade = '';
			break;
	}

	return { grade, rules };
}
